torch/csrc/api/include/torch/data/transforms/collate.h

```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+#include <torch/data/example.h>
+#include <torch/data/transforms/base.h>
+#include <torch/data/transforms/lambda.h>
+#include <torch/tensor.h>
+
+#include <functional>
+#include <utility>
+#include <vector>
+
+namespace torch {
+namespace data {
+namespace transforms {
+
+template <typename T, typename BatchType = std::vector<T>>
+using Collation = BatchTransform<BatchType, T>;
+
+// A `Collate` allows passing a custom function to reduce/collate a batch
+// into a single value.
+template <typename T, typename BatchType = std::vector<T>>
+using Collate = BatchLambda<BatchType, T>;
+
+template <typename E = Example<>>
+struct Stack;
+
+template <>
+struct Stack<Example<>> : public Collation<Example<>> {
+  Example<> apply_batch(std::vector<Example<>> examples) override {
+    std::vector<torch::Tensor> data, labels;
+    data.reserve(examples.size());
+    labels.reserve(examples.size());
+    for (auto& example : examples) {
+      data.push_back(std::move(example.data));
+      labels.push_back(std::move(example.label));
+    }
+    return {torch::stack(data), torch::stack(labels)};
+  }
+};
+
+template <>
+struct Stack<Tensor> : public Collation<Tensor> {
+  Tensor apply_batch(std::vector<Tensor> tensors) override {
+    return torch::stack(tensors);
+  }
+};
```

Comment:
It would be nice if we could have a bunch of default instances for a standard definition of `Collate` (like we do in Python). Sort of how `std::hash` already works with most common types.

Reply:
Good idea. So like `Tensor`, and then numeric types?

Reply:
Yeah, check the Python collating function for inspiration